### PR TITLE
Converting KMeans++trainer to estimator.

### PR DESCRIPTION
**Pull request by sfilipi**
Ongoing work to address #754

### CODE REVIEW

File: test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs [hunk 1]
```diff
@@ -16,5 +17,34 @@ public partial class TrainerEstimators : TestDataPipeBase
         public TrainerEstimators(ITestOutputHelper helper) : base(helper)
         {
         }
+
+        /// <summary>
+        /// FastTreeBinaryClassification TrainerEstimator test
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs]: >FastTreeBinaryClassification [](start = 12, length = 28)

fix comment

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 2]
```diff
@@ -203,6 +233,18 @@ private static int ComputeNumThreads(IHost host, int? argNumThreads)
             return Math.Max(1, maxThreads);
         }
 
+        private static SchemaShape.Column MakeWeightColumn(string weightColumn)
```
Comment (sfilipi) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >MakeWeightColumn [](start = 42, length = 16)

re-use from TrainerUtils

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs [hunk 3]
```diff
@@ -421,6 +428,65 @@ private static VersionInfo GetVersionInfo()
         }
     }
 
+    /// <summary>
+    /// Base class for the <see cref="ISingleFeaturePredictionTransformer{TModel}"/> working on clustering tasks.
+    /// </summary>
+    /// <typeparam name="TModel">An implementation of the <see cref="IPredictorProducing{TResult}"/></typeparam>
+    public sealed class ClusteringPredictionTransformer<TModel> : SingleFeaturePredictionTransformerBase<TModel>
```
Comment (sfilipi) [inline src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: >ClusteringPredictionTransformer [](start = 24, length = 31)

I still don't like not making those generic on the scorer as well...

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 4]
```diff
@@ -17,6 +17,7 @@
 using Microsoft.ML.Runtime.Training;
 using Microsoft.ML.Runtime.Internal.Internallearn;
 using Microsoft.ML.Runtime.EntryPoints;
+using Microsoft.ML.Core.Data;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: sort order

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs [hunk 5]
```diff
@@ -421,6 +428,65 @@ private static VersionInfo GetVersionInfo()
...
+        }
+
+        protected override void SaveCore(ModelSaveContext ctx)
+        {
+            Contracts.AssertValue(ctx);
+            ctx.SetVersionInfo(GetVersionInfo());
+
+            // *** Binary format ***
+            // <base info>
+            // id of string: scorer threshold column
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: > // id of string: scorer threshold column [](start = 11, length = 41)

doesn't look right

File: src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs (hunk 3 above)
Comment (sfilipi) [reply src/Microsoft.ML.Data/Scorers/PredictionTransformer.cs]: addressed on #996

---
In reply to: [219580106](https://github.com/dotnet/machinelearning/pull/979#discussion_r219580106) [](ancestors = 219580106)

File: test/Microsoft.ML.TestFramework/Datasets.cs [hunk 6]
```diff
@@ -600,15 +600,15 @@ public static class TestDatasets
...
+            trainFilename = @"Train-28x28.txt",
+            testFilename = @"Test-28x28.txt"
         };
 
         public static TestDataset mnistTiny28 = new TestDataset()
         {
             name = "mnistTiny28",
-            trainFilename = @"..\MNIST\Train-Tiny-28x28.txt",
-            testFilename = @"..\MNIST\Test-Tiny-28x28.txt"
+            trainFilename = @"Train-Tiny-28x28.txt",
```
Comment (Zruty0) [inline test/Microsoft.ML.TestFramework/Datasets.cs]: >Train [](start = 30, length = 5)

intentional

File: test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs [hunk 7]
```diff
@@ -17,6 +18,35 @@ public TrainerEstimators(ITestOutputHelper helper) : base(helper)
...
+            });
+            var data = reader.Read(new MultiFileSource(GetDataPath(TestDatasets.mnistTiny28.trainFilename)));
+
+
+            // Pipeline.
+            var pipeline = new KMeansPlusPlusTrainer(Env, featureColumn, weightColumn,
+                            advancedSettings: s => { s.InitAlgorithm = KMeansPlusPlusTrainer.InitAlgorithm.KMeansParallel; });
+
+            TestEstimatorCore(pipeline, data);
+        }
```
Comment (Zruty0) [inline test/Microsoft.ML.Tests/TrainerEstimators/TrainerEstimators.cs]: >} [](start = 8, length = 1)

Call `Done()` at the end

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 8]
```diff
@@ -76,21 +75,50 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
...
         public override PredictionKind PredictionKind => PredictionKind.Clustering;
 
-        public KMeansPlusPlusTrainer(IHostEnvironment env, Arguments args)
-            : base(env, LoadNameValue)
+        /// <summary>
+        /// Initializes a new instance of <see cref="KMeansPlusPlusTrainer"/>
+        /// </summary>
+        /// <param name="env">The private instance of <see cref="IHostEnvironment"/>.</param>
+        /// <param name="featureColumn">The name of the feature column.</param>
+        /// <param name="weightColumn">The name for the column containing the initial weight.</param>
```
Comment (Zruty0) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >initial [](start = 78, length = 7)

example weights, not initial weight

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 9]
```diff
@@ -76,21 +75,50 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
...
-            Host.CheckValue(args, nameof(args));
+            if (args == null)
+                args = new Arguments();
+
+            if (advancedSettings != null)
+                advancedSettings.Invoke(args);
+
             Host.CheckUserArg(args.K > 0, nameof(args.K), "Must be positive");
 
+            // is this even necessary, if there is only one column, for example
```
Comment (Zruty0) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: >is this even necessary, if there is only one column, for example [](start = 15, length = 64)

It checks for non-emptiness of string

**Comment by Zruty0** [comment]
This calls for a `ClusteringContext` with a pigstension for KMeans

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs [hunk 10]
```diff
@@ -76,21 +75,50 @@ public class Arguments : UnsupervisedLearnerInputBaseWithWeight
...
+        {
+
+        }
+
+        private KMeansPlusPlusTrainer(IHostEnvironment env, Arguments args, string featureColumn, string weightColumn, Action<Arguments> advancedSettings = null)
+            : base(Contracts.CheckRef(env, nameof(env)).Register(LoadNameValue), TrainerUtils.MakeR4VecFeature(featureColumn), null, TrainerUtils.MakeR4ScalarWeightColumn(weightColumn))
         {
-            Host.CheckValue(args, nameof(args));
+            if (args == null)
+                args = new Arguments();
```
Comment (TomFinley) [inline src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: The net effect of this change is that the internal constructor is now tolerant to passing in `null` arguments.. Even though the constructor is internal I'd prefer to keep it clean, and maintain the invariants we have everywhere else. Note the `Host.CheckValue(args, nameof(args));` check that was deleted.

If you still had the check, and just passed in `new Arguments` in the above constructor, that would have had no more lines of code.

**Comment by TomFinley** [comment]
> using Microsoft.ML.Runtime;

I like how when you changed from `Float` to `float` you made those two separate commits. That makes things easier to check out.

---
Refers to: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs:10 in bd436bb. [](commit_id = bd436bb616fd9aadf8190f6fc314f42332684202, deletion_comment = True)

File: test/Microsoft.ML.TestFramework/Datasets.cs (hunk 6 above)
Comment (sfilipi) [reply test/Microsoft.ML.TestFramework/Datasets.cs]: yes. The 'old' path is from the TLC solution structure. 

---
In reply to: [219997779](https://github.com/dotnet/machinelearning/pull/979#discussion_r219997779) [](ancestors = 219997779)

**Comment by sfilipi** [comment]
> using Microsoft.ML.Runtime;

thanks for pointing out. Will keep that in mind.

---
In reply to: [424419588](https://github.com/dotnet/machinelearning/pull/979#issuecomment-424419588) [](ancestors = 424419588)

---
Refers to: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs:10 in bd436bb. [](commit_id = bd436bb616fd9aadf8190f6fc314f42332684202, deletion_comment = True)

File: src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs (hunk 10 above)
Comment (sfilipi) [reply src/Microsoft.ML.KMeansClustering/KMeansPlusPlusTrainer.cs]: appreciate pointing this out. 

---
In reply to: [220271405](https://github.com/dotnet/machinelearning/pull/979#discussion_r220271405) [](ancestors = 220271405)

File: src/Microsoft.ML.Data/Evaluators/ClusteringEvaluator.cs [hunk 11]
```diff
@@ -52,6 +52,52 @@ public ClusteringEvaluator(IHostEnvironment env, Arguments args)
...
+        /// <summary>
+        /// Evaluates scored clustering data.
+        /// </summary>
+        /// <param name="data">The scored data.</param>
+        /// <param name="label">The name of the label column in <paramref name="data"/>.</param>
+        /// <param name="score">The name of the score column in <paramref name="data"/>.</param>
+        /// <param name="predictedLabel">The name of the predicted label column in <paramref name="data"/>.</param>
+        /// <param name="features">The name of the feature column in <paramref name="data"/>.</param>
+        /// <returns>The evaluation results.</returns>
+        public Result Evaluate(IDataView data, string label, string score, string predictedLabel, string features = null)
```
[the remark this answers is not in the document]
Comment (Zruty0) [reply src/Microsoft.ML.Data/Evaluators/ClusteringEvaluator.cs]: OK, I double-checked. `label` is required for NMI, if it's not provided, NMI is not calculated. That makes sense. Let's make it optional.
`predictedLabel` is never required, as I suspected. Let's remove it.

---
In reply to: [220990067](https://github.com/dotnet/machinelearning/pull/979#discussion_r220990067) [](ancestors = 220990067)
